### PR TITLE
[device manager] fix panic of nil pointer dereference in UT TestNewManagerImplStartProbeMode

### LINKED ISSUE #121451: [Flake] unit test TestNewManagerImplStartProbeMode
Seen in https://github.com/kubernetes/kubernetes/pull/121445#issuecomment-1775347291 and in https://github.com/kubernetes/kubernetes/pull/121407

No hits on https://storage.googleapis.com/k8s-triage/index.html?pr=1&test=TestNewManagerImplStartProbeMode

stack strace

```
E1023 13:40:22.589536   43369 goroutinemap.go:150] Operation for "/tmp/device_plugin1067413082/server.sock" failed. No retries permitted until 2023-10-23 13:40:23.0894569 +0000 UTC m=+0.753948700 (durationBeforeRetry 500ms). Error: RegisterPlugin error -- failed to get plugin info using RPC GetInfo at socket /tmp/device_plugin1067413082/server.sock, err: rpc error: code = Unimplemented desc = unknown service pluginregistration.Registration
--- FAIL: TestNewManagerImplStartProbeMode (0.01s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x28fb0ac]

goroutine 238 [running]:
testing.tRunner.func1.2({0x2b14fa0, 0x452d660})
	/usr/local/go/src/testing/testing.go:1545 +0x366
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1548 +0x630
panic({0x2b14fa0?, 0x452d660?})
	/usr/local/go/src/runtime/panic.go:920 +0x270
k8s.io/kubernetes/pkg/kubelet/cm/devicemanager.(*ManagerImpl).PluginDisconnected(0xc00068fd40, {0xc000157470, 0x14})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/manager.go:244 +0x46c
k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1.(*client).Disconnect(0xc0000d7ae0)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go:106 +0x2d9
k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1.(*server).disconnectClient(0xc000714270?, {0xc000157470, 0x14}, {0x3232508, 0xc0000d7ae0})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1/handler.go:86 +0x55
k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1.(*server).Stop.func1({0xc000157470, 0x14}, {0x3232508, 0xc0000d7ae0})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1/server.go:120 +0x7a
k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1.(*server).visitClients(0xc000374d90, 0xc000763d38)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1/server.go:186 +0x11f
k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1.(*server).Stop(0xc000374d90)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1/server.go:119 +0x73
k8s.io/kubernetes/pkg/kubelet/cm/devicemanager.(*ManagerImpl).Stop(...)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/manager.go:315
k8s.io/kubernetes/pkg/kubelet/cm/devicemanager.cleanup(0x322a6f0?, {0x3251ca0, 0xc0006fb480}, 0x0?)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/manager_test.go:335 +0x42
k8s.io/kubernetes/pkg/kubelet/cm/devicemanager.TestNewManagerImplStartProbeMode(0x0?)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/manager_test.go:119 +0x126
testing.tRunner(0xc0007081a0, 0x2ef5a08)
	/usr/local/go/src/testing/testing.go:1595 +0x239
```

nothing evident on the git history of the test itself, but it seems @pohly [added it to detect races](https://github.com/kubernetes/kubernetes/commit/fcbb64b93dfb633f880bad9ed1771201843b8e7e), so most probable something in the device manager has changed?

/sig node
/kind flake

### PR DESCRIPTION
**Pull request by wzshiming**
#### What type of PR is this?

/kind flake

#### Which issue(s) this PR fixes:

Fixes #121451

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

**Comment by pohly** [comment]
Can you explain why this fixes the flake? Why does it occur, why is this the right solution?

**Comment by aojea** [comment]
@wzshiming can you please run the test with  `stress`https://gist.github.com/liggitt/6a3a2217fa5f846b52519acfc0ffece0#running-unit-tests-to-reproduce-flakes and paste the output on the description for reference

**Comment by aojea** [comment]
why we notice this only recently? some new feature graduated that exercise this path now?

**Comment by aojea** [comment]
This seems to solve the panic
```
 go test -timeout 30s -run ^TestNewManagerImplStartProbeMode$ k8s.io/kubernetes/pkg/kubelet/cm/devicemanager -v -count 1 -c
aojea@aojea:~/src/kubernetes$ stress ./devicemanager.test -test.run TestNewManagerImplStartProbeMode
5s: 2363 runs so far, 0 failures
10s: 4726 runs so far, 0 failures
15s: 7067 runs so far, 0 failures
20s: 9382 runs so far, 0 failures
25s: 11688 runs so far, 0 failures
30s: 14003 runs so far, 0 failures
35s: 16338 runs so far, 0 failures
40s: 18617 runs so far, 0 failures
45s: 20857 runs so far, 0 failures
```

however, if run with the race detector

```
/tmp/go-stress-20231024T140936-4196658318
I1024 14:09:39.722379  871929 fake_topology_manager.go:33] "NewFakeManager"
I1024 14:09:39.722694  871929 manager.go:471] "Failed to read data from checkpoint" checkpoint="kubelet_internal_checkpoint" err="checkpoint is not found"
I1024 14:09:39.723269  871929 stub.go:135] "Starting device plugin server"
I1024 14:09:39.726268  871929 stub.go:176] "Starting to serve on socket" socket="/tmp/device_plugin713883419/device-plugin.sock"
I1024 14:09:39.726341  871929 stub.go:196] "Stopping device plugin server"
I1024 14:09:39.726901  871929 plugin_manager.go:118] "Starting Kubelet Plugin Manager"
I1024 14:09:39.736062  871929 stub.go:253] "GetInfo"
E1024 14:09:39.740970  871929 goroutinemap.go:150] Operation for "/tmp/device_plugin713883419/server.sock" failed. No retries permitted until 2023-10-24 14:09:40.240884678 +0000 UTC m=+0.668252748 (durationBeforeRetry 500ms). Error: RegisterPlugin error -- failed to get plugin info using RPC GetInfo at socket /tmp/device_plugin713883419/server.sock, err: rpc error: code = Unimplemented desc = unknown service pluginregistration.Registration
==================
WARNING: DATA RACE
Read at 0x00c00052e9d8 by goroutine 140:
  k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1.(*client).Disconnect()
      /usr/local/google/home/aojea/src/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go:99 +0x59
  k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1.(*server).disconnectClient()
      /usr/local/google/home/aojea/src/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1/handler.go:86 +0x54
  k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1.(*server).Stop.func1()
      /usr/local/google/home/aojea/src/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1/server.go:120 +0x79
  k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1.(*server).visitClients()
      /usr/local/google/home/aojea/src/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1/server.go:186 +0x11e
  k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1
…

/tmp/go-stress-20231024T140936-2968540325
```

**Comment by wzshiming** [comment]
ohh, Sorry to see this so late...

**Comment by ffromani** [comment]
/triage accepted
/priority important-longterm

the last time this code changed significantly was in db88676c2062
I can't remember any recent changes which exercise the plugin disconnection flow more. Testing in this area can use some love, though.

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 1]
```diff
@@ -236,12 +236,12 @@ func (m *ManagerImpl) PluginDisconnected(resourceName string) {
...
 
-	if _, exists := m.endpoints[resourceName]; exists {
+	if ep, exists := m.endpoints[resourceName]; exists {
 		m.markResourceUnhealthy(resourceName)
-		klog.V(2).InfoS("Endpoint became unhealthy", "resourceName", resourceName, "endpoint", m.endpoints[resourceName])
-	}
+		klog.V(2).InfoS("Endpoint became unhealthy", "resourceName", resourceName, "endpoint", ep)
 
-	m.endpoints[resourceName].e.setStopTime(time.Now())
+		ep.e.setStopTime(time.Now())
```
[the remark this answers is not in the document]
Comment (ffromani) [reply pkg/kubelet/cm/devicemanager/manager.go:243]: the fix per se makes sense to me, but I'm not clear on the sequence of events which triggers `PluginDisconnected` with missing endpoint for a resource.

I tend to believe this fix is safe and worthy anyway, in general.

**Review by ffromani — COMMENTED** [review]
I like to understand the sequence of events better, but the fix makes sense in general

**Comment by pacoxu** [comment]
/retitle [device manager] fix panic of nil pointer dereference in UT TestNewManagerImplStartProbeMode

File: pkg/kubelet/cm/devicemanager/manager.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (aojea) [reply pkg/kubelet/cm/devicemanager/manager.go:243]: I think the bug is very clear

**Comment by aojea** [comment]
ok, found the other race in the client, it need to lock sent https://github.com/kubernetes/kubernetes/pull/121494 keeping hist commit because most probably @wzshiming is in his night

**Pull request closed without merging**